WorkerPool/WorkerPool_test.go

```diff
@@ -0,0 +1,678 @@
+		},
+	}
+
+	for _, testCase := range testCases {
+		testAlias := testCase.TestAlias
+		initWorkerNumber := testCase.InitWorkerNumber
+		startWorkerNumber := testCase.StartWorkerNumber
+		expectedStartedWorkers := testCase.ExpectedStartedWorkers
+
+		fn := func(t *testing.T) {
+
+			// chanel workers are waiting for close
+			block := make(chan struct{})
+
+			dep := NewWorkerPool(initWorkerNumber)
+
+			actualWorkerNumber := 0
+			mtx := sync.RWMutex{}
+
+			timeOut := time.Duration(time.Millisecond)
+
+			for i := 0; i < startWorkerNumber; i++ {
+				_ = dep.Do(func() {
+					mtx.Lock()
+					actualWorkerNumber++
+					mtx.Unlock()
+					<-block
+				}, timeOut)
+			}
+
+			_ = <-time.After(timeOut)
+
+			mtx.RLock()
+			actualStartedWorkers := actualWorkerNumber
+			mtx.RUnlock()
+
+			// release workers
+			defer close(block)
+
+			if actualStartedWorkers != expectedStartedWorkers {
+				t.Errorf("For TestAlias '%s' WorkerPool.Do()  \r\n started %v workers \r\n while expected %v \r\n", testAlias, actualStartedWorkers, expectedStartedWorkers)
+			}
+		}
+		t.Run(testAlias, fn)
+	}
+
+}
+
+func Test_WorkerPoolDoProcessAllWorkers(t *testing.T) {
+
+	testCases := []struct {
+		TestAlias            string
+		InitWorkerNumber     int
+		StartWorkerNumber    int
+		ExpectedDonedWorkers int
+	}{
+		{
+			TestAlias:            "2 work items for 0 workers",
+			InitWorkerNumber:     0,
+			StartWorkerNumber:    2,
+			ExpectedDonedWorkers: 0,
+		},
+		{
+			TestAlias:            "1 work item for 1 worker",
+			InitWorkerNumber:     1,
+			StartWorkerNumber:    1,
+			ExpectedDonedWorkers: 1,
+		},
+		{
+			TestAlias:            "1 work item for 10 workers",
+			InitWorkerNumber:     10,
+			StartWorkerNumber:    1,
+			ExpectedDonedWorkers: 1,
+		},
+		{
+			TestAlias:            "20 work items for 16 workers",
+			InitWorkerNumber:     16,
+			StartWorkerNumber:    20,
+			ExpectedDonedWorkers: 20,
+		},
+		{
+			TestAlias:            "1000 work items for 16 workers",
+			InitWorkerNumber:     16,
+			StartWorkerNumber:    1000,
+			ExpectedDonedWorkers: 1000,
+		},
+		{
+			TestAlias:            "1000 work items for 1 workers",
+			InitWorkerNumber:     1,
+			StartWorkerNumber:    1000,
+			ExpectedDonedWorkers: 1000,
+		},
+	}
+
+	for _, testCase := range testCases {
+		testAlias := testCase.TestAlias
+		initWorkerNumber := testCase.InitWorkerNumber
+		startWorkerNumber := testCase.StartWorkerNumber
+		expectedDoneWorkers := testCase.ExpectedDonedWorkers
+
+		fn := func(t *testing.T) {
+
+			dep := NewWorkerPool(initWorkerNumber)
+
+			actualWorkerDone := 0
+			mtx := sync.RWMutex{}
+
+			// timeout should be long enough not to happen
+			timeOut := time.Duration(time.Millisecond)
+
+			wg := sync.WaitGroup{}
+
+			for i := 0; i < startWorkerNumber; i++ {
+				wg.Add(1)
+				err := dep.Do(func() {
+					mtx.Lock()
+					actualWorkerDone++
+					mtx.Unlock()
+					wg.Done()
+				}, timeOut)
+				if err != nil {
+					wg.Done()
```

Comment:
To be 100%% clear, it should be wrapped into local anonymous function and `defer wg.Done()`. However, actually there is a bug in testing, which invokes `defer` in a very specific way.